### PR TITLE
Rebase onto latest upstream

### PR DESCRIPTION
**Pull request by zml2008**
This is a fairly straightforward rebase based on what jetbrains has changed.

Some key points:

- Upstream builds against Java 11. There are a few minor changes needed to bring this back to Java 8 - those are in the first patch.
- Some restructuring occurred so `StructMethod`s no longer hold a reference to their owning `StructClass` -- there is no StructClass available at method construction time, but the qualified name is available so I used that instead. The only place where that is awkward is [`master...zml2008:rebase`#diff-da12830b32](https://github.com/MinecraftForge/ForgeFlower/compare/master...zml2008:rebase#diff-da12830b32e0ccf4077250c70c5e39c1b571757239deedb68f3708753230f72fR1754)
- There is a forgeflower patch for module and package-info constant pool handling combined with debug logging. Upstream's support for those files makes only the debug logging part of that patch relevant -- it may want to be renamed?

These seem to have reordered one local variable declaration -- not sure why (output of MCPConfig's `testDecompileJoinedHotspot8` on 1.16.5):

```diff
diff --git a/net/minecraft/client/resources/PackResourcesAdapterV4.java b/net/minecraft/client/resources/PackResourcesAdapterV4.java
index a5c3444..80b06dd 100644
--- a/net/minecraft/client/resources/PackResourcesAdapterV4.java
+++ b/net/minecraft/client/resources/PackResourcesAdapterV4.java
@@ -148,8 +148,8 @@ public class PackResourcesAdapterV4 implements PackResources {
    public static InputStream m_118773_(InputStream p_118774_, InputStream p_118775_, int p_118776_, int p_118777_, int p_118778_, int p_118779_, int p_118780_) throws IOException {
       ByteArrayInputStream bytearrayinputstream;
       try (
-         NativeImage nativeimage1 = NativeImage.m_85058_(p_118775_);
          NativeImage nativeimage = NativeImage.m_85058_(p_118774_);
+         NativeImage nativeimage1 = NativeImage.m_85058_(p_118775_);
       ) {
          int i = nativeimage.m_84982_();
          int j = nativeimage.m_85084_();
```

I've confirmed with Sponge's toolchain that the `package-info` files added in 21w13a decompile correctly as well.

### CODE REVIEW

**Review by TheCurle — APPROVED** [review]
Some _interesting_ solutions to some simple problems, but seems to work fine in preliminary testing.

I tried doing this and failed miserably, so the fact it merely compiles is good enough for me, let alone that MCPConfig accepts this for an input.

File: FernFlower-Patches/0001-Git-filter-and-setup-backport-to-Java-8.patch [hunk 1]
```diff
@@ -2443,3 +2442,34 @@ index f9553162f122c71b34635112e717c3e733b5b212..62bd9b9ccefea2b65ae41e5d9a545e20
...
+diff --git a/src/org/jetbrains/java/decompiler/util/SFormsFastMapDirect.java b/src/org/jetbrains/java/decompiler/util/SFormsFastMapDirect.java
+index bb8b4dd9b866ec54491da50310b09b6fcb0bdf3f..7143281598c9273e847cc8aa5d86178c3b2b6bab 100644
+--- a/src/org/jetbrains/java/decompiler/util/SFormsFastMapDirect.java
++++ b/src/org/jetbrains/java/decompiler/util/SFormsFastMapDirect.java
+@@ -342,7 +342,7 @@ public class SFormsFastMapDirect {
+         if (ent != null) {
+           final int key = i == 0 ? ikey : (i == 1 ? ikey + VarExprent.STACK_BASE : -ikey);
+ 
+-          list.add(new Entry<>() {
++          list.add(new Entry<Integer, FastSparseSet<Integer>>() {
```
Comment (LexManos) [inline FernFlower-Patches/0001-Git-filter-and-setup-backport-to-Java-8.patch:2472]: This line just reverts a upstream change.

Comment (zml2008) [reply FernFlower-Patches/0001-Git-filter-and-setup-backport-to-Java-8.patch:2472]: correct -- the upstream change is not valid on Java 8

**Comment by LexManos** [comment]
> it may want to be renamed?
Feel free. They are commit messages so being descriptive is nice.

> These seem to have reordered one local variable declaration 
Ya, I've seen this one change before, I don't remember ever being able to figure out why... Is that the only diff between the full decompile?
Also, anyone run this through the JVM tests in MCPConfig yet?

File: FernFlower-Patches/0001-Git-filter-and-setup-backport-to-Java-8.patch (hunk 1 above)
Comment (LexManos) [reply FernFlower-Patches/0001-Git-filter-and-setup-backport-to-Java-8.patch:2472]: Ah.. well that makes sense then.

**Comment by zml2008** [comment]
updated -- that's the only diff in a full decompile, yeah.

i ran through the `testDecompileJoined` task in MCPConfig and nothing seemed to blow up, but only did the full diff on the J8 decompile.

**Comment by JDLogic** [comment]
> These seem to have reordered one local variable declaration

The current FF version used in MCPConfig does not include this commit (9c8685efbd72dc38a945b60f4cc899d1a3d99054) which is causing the diff.

**Comment by zml2008** [comment]
ah, that'd do it -- thanks!

**Comment by LexManos** [comment]
Just a note, the main point of this bump is to update to a version of FF that support package-info classes properly.
Which changes the decompiled package infos that now exist in 21w13a from:
```java
package net.minecraft;

import javax.annotation.ParametersAreNonnullByDefault;

// $FF: synthetic class
@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
@FieldsAreNonnullByDefault
interface package-info {
}
```
To the proper:
```java
@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
@FieldsAreNonnullByDefault
package net.minecraft;

import javax.annotation.ParametersAreNonnullByDefault;
```

Ran it through the decompiler tests on MCPConfig {Adopt HotSpot 8/11/15} and it seems to work fine. 
Gunna pull this locally, so I can tag it correctly before publishing.

**Pull request merged**